### DISCUSSION TITLE
collapsing undo boundary

### DISCUSSION THREAD

**Discussion by jgarte** [meow-edit/meow#446] [Q&A]
> [and tries to improve the experience by collapsing undo boundary](https://github.com/meow-edit/meow/blob/master/EXPLANATION.org)

Hi, what is meant  by "collapsing undo boundary" exactly?

 Can this notion/concept use some more explanation/disambiguation when it is introduced in the context of `EXPLANATION.org`?

I couldn't make it out...

**Comment by DogLooksGood** [comment]
Usually after a kmacro execution, you have to undo many times to get back to the original state. In beacon state, meow collapses all execution into one undo boundary, so you can undo the effect with just one undo. However, for the changes you made when recording the kmacro, you have to undo multiple times as usual, as it should be.

The undo boundary is a concept in Emacs, so we don't pull it into the explanation.

**Reply by jgarte** [reply]
> The undo boundary is a concept in Emacs, so we don't pull it into the explanation.

I think that it might still be helpful to introduce/mention that `undo-boundary` is an Emacs related concept for the reader. They might not know that undo-boundary is [some function in Emacs](https://www.gnu.org/software/emacs/manual/html_node/elisp/Undo.html#index-undo_002dboundary) and will be confused.

**Reply by Ziqi-Yang** [reply]
But I see I still have to undo many times to return to the original state.   
I test it with a nearly bear setup of `emacs 29` with meow:
```
1 2 3
```
In beacon state, I use `b` to select all numbers and type `i` to insert something around number `3`, then I use `esc` to exit the insert state. Then ( for both situation whether quit beacon state or not) I still need to undo many times to return to the original state.
